### PR TITLE
Cmr 8457 generic associations feature new

### PR DESCRIPTION
**Pull request by sxu123**

### CODE REVIEW

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 1]
```diff
@@ -407,6 +414,7 @@
 
 (defmethod index-concept :default
   [context concept parsed-concept options]
+  ;;(println "In index-concept :default for: " (:concept-id concept))
   (let [{:keys [all-revisions-index?]} options
```
Comment (eereiter) [inline indexer-app/src/cmr/indexer/services/index_service.clj:418]: remove the println.

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 2]
```diff
@@ -439,6 +449,10 @@
                                       context service-associations)
                 tool-associations (es/parse-non-tombstone-associations
                                    context tool-associations)
+                ;;_ (println "generic-associations before removing tombstone: " generic-associations)
+                generic-associations (es/parse-non-tombstone-associations
```
Comment (eereiter) [inline indexer-app/src/cmr/indexer/services/index_service.clj:453]: remove the println

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 3]
```diff
@@ -439,6 +449,10 @@
...
+                                      context generic-associations)
+                ;;_ (println "generic-associations after removing tombstone: " generic-associations)
```
Comment (eereiter) [inline indexer-app/src/cmr/indexer/services/index_service.clj:455]: Remove the println

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 4]
```diff
@@ -489,6 +504,36 @@
         parsed-concept (cp/parse-concept context concept)]
     (index-concept context concept parsed-concept options)))
 
+(defn- index-associated-generic-concept
+  "Given a concept id, revision id, index the concept to which it refers."
+  [context concept-id revision-id options]
+  ;;(println "index concept id: " concept-id)
```
Comment (eereiter) [inline indexer-app/src/cmr/indexer/services/index_service.clj:510]: remove the println

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 5]
```diff
@@ -489,6 +504,36 @@
...
+  (let [concept (if revision-id
+                  (meta-db/get-concept context concept-id revision-id)
+                  (meta-db/get-latest-concept context concept-id))
+        parsed-concept (cp/parse-concept context concept)]
+    (index-concept context concept parsed-concept options)))
+
+(defn- index-associated-generic-source
+  "Index the associated source concept of the given generic association concept."
+  [context concept options]
+  ;;(println "index-associated-generic-source")
```
Comment (eereiter) [inline indexer-app/src/cmr/indexer/services/index_service.clj:520]: remove the println

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 6]
```diff
@@ -489,6 +504,36 @@
...
+  (index-associated-generic-concept
+   context
+   (get-in concept [:extra-fields :source-concept-identifier])
+   (get-in concept [:extra-fields :source-revision-id])
+   options))
+
+(defn- index-associated-generic-destination
+  "Index the associated destination concept of the given generic association concept."
+  [context concept options]
+  ;;(println "index-associated-generic-destination")
```
Comment (eereiter) [inline indexer-app/src/cmr/indexer/services/index_service.clj:530]: remove the println

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 7]
```diff
@@ -522,14 +567,19 @@
   [context concept parsed-concept options]
   (index-associated-collection context concept options))
 
+(defmethod index-concept :generic-association
+  [context concept parsed-concept options]
+  ;;(println "In index-concept :generic-association: " concept)
```
Comment (eereiter) [inline indexer-app/src/cmr/indexer/services/index_service.clj:572]: remove the println

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 8]
```diff
@@ -631,6 +681,13 @@
   ;; This is the same thing we do when a tool association is updated. So we call the same function.
   (index-association-concept context concept-id revision-id options))
 
+(defmethod delete-concept :generic-association
+  [context concept-id revision-id options]
+  (let [concept (meta-db/get-concept context concept-id revision-id)]
+    ;;(println "In delete-concept :generic-association: " concept)
```
Comment (eereiter) [inline indexer-app/src/cmr/indexer/services/index_service.clj:687]: remove the println

**Review by eereiter — APPROVED** [review]

File: search-app/src/cmr/search/services/generic_association_service.clj [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        ;; Need to set the original revision-id back so that it can be inserted into
+        ;; db correctly.
+        concept (assoc concept :revision-id revision-id)
+        associations (assoc-validation/associations-json->associations associations-json)
+        associations (assoc-validation/validate-no-same-concept-generic-association
+                      concept
+                      associations)
+        associations (assoc-validation/validate-generic-association-types
+                      concept
+                      associations)
```
Comment (zimzoom) [inline search-app/src/cmr/search/services/generic_association_service.clj:241]: I believe you can combine lines 235-241 with the 'last arg' threading operator ->>, like this:

(->> associations-json
(assoc-validation/associations-json->associations)
(assoc-validation/validate-no-same-concept-generic-association concept)
(assoc-validation/validate-generic-association-types concept))

**Review by zimzoom — APPROVED** [review]

File: indexer-app/src/cmr/indexer/data/concepts/generic.clj [hunk 10]
```diff
@@ -63,11 +71,17 @@
     {(keyword field-name) value
      (keyword field-name-lower) value-lower}))
 
-(defn- parsed-concept->elastic-doc-without-context
-  "Generate an all of the elastic document parts that do not require a context"
-  [concept parsed-concept]
+(defn- parsed-concept->elastic-doc
+  "Generate an all of the elastic document parts"
```
[the remark this answers is not in the document]
Comment (sxu123) [reply indexer-app/src/cmr/indexer/data/concepts/generic.clj:75]: Strange phrase. It just means to generate an elastic document. Again, it's not easy to get a green build, so I'm gonna take a note to fix this in the next ticket.

**Pull request merged**